Hakawai/HKWMentionsPluginV2.m

```diff
@@ -0,0 +1,1590 @@
+@property (nonatomic, strong) id<HKWMentionsCreationStateMachine> creationStateMachine;
+
+/*!
+ The point at which the last character typed was inserted.
+ */
+@property (nonatomic) NSUInteger previousInsertionLocation;
+
+/*!
+ The previous distinct value of the parent text view's \c selectedRange value for which the text view was in insertion
+ mode.
+ */
+@property (nonatomic) NSRange previousSelectionRange;
+@property (nonatomic) NSUInteger previousTextLength;
+
+/*!
+ The mention currently highlighted as the 'selected' mention (due to the cursor being in the right place), or the
+ mention which may be selected if the state machine is in the 'about to select mention' state.
+ */
+@property (nonatomic, strong) HKWMentionsAttribute *currentlySelectedMention;
+
+/*!
+ The range of the mention attribute whose value is stored in the \c currentlySelectedMention property. This value is not
+ automatically updated; it must be kept in sync by the plug-in if the text is modified.
+ */
+@property (nonatomic) NSRange currentlySelectedMentionRange;
+
+@property (nonatomic, strong) NSDictionary *mentionSelectedAttributes;
+@property (nonatomic, strong) NSDictionary *mentionUnselectedAttributes;
+
+@property (nonatomic, readwrite) HKWMentionsChooserPositionMode chooserPositionMode;
+
+@property (nonatomic, readonly) BOOL viewportLocksToTopUponMentionCreation;
+@property (nonatomic, readonly) BOOL viewportLocksToBottomUponMentionCreation;
+@property (nonatomic, readonly) BOOL viewportLocksUponMentionCreation;
+
+/*!
+ If the state machine is currently executing the \c advanceStateForCharacterInsertion method, this property contains the
+ character being inserted. Otherwise, it contains the NULL character, 0.
+
+ This property is intended to allow the data-retrieved callback code to determine whether the callback was called
+ synchronously or asynchronously, and therefore determine which character to use to determine whether mentions creation,
+ if cancelled, should be allowed to immediately restart.
+
+ (If the callback was made asynchronously, then the text in the text view should be used to determine the last character
+ typed. If the callback was made synchronously, then the character in this property should be used to determine whether
+ or not to allow immediate restart. However, sometimes the callback will be deferred until the next runloop anyways, so
+ this should not be used to determine whether or not the callback is synchronous.)
+ */
+@property (nonatomic) unichar characterForAdvanceStateForCharacterInsertion;
+
+// Properties for resuming mentions creation
+@property (nonatomic) BOOL shouldResumeMentionsCreation;
+@property (nonatomic) unichar resumeMentionsControlCharacter;   // 0 if implicit mention
+@property (nonatomic) NSUInteger resumeMentionsPriorTextLength;
+@property (nonatomic) NSUInteger resumeMentionsPriorPosition;
+@property (nonatomic) NSString *resumeMentionsPriorString;
```

Comment:
can you double check if all those properties are actually being used?

Reply:
yes removed